doc/design/refactor/distributed_architecture.md

```diff
@@ -136,18 +136,43 @@ iteratively.
 
 As shown in the graph, `RemoteExecutor.run` sends the IR to the
 PaddlePaddle cluster for Execution. You can also use parameter
-`fetch_list` to interactively fetch varirable back to local for
+`fetch_list` to interactively fetch variable back to local for
 log printing.
 
 The Python `RemoteExecutor` is derived from `Executor` class.
 For more information about `RemoteExecutor`, please
 see [Design Doc: RemoteExecutor](./remote_executor.md).
 
+The `RemoteExecutor.run` interface defination is:
+
+```python
+run(self,
+    program=None,
+    feed=None,
+    fetch_list=None,
+    feed_var_name='feed',
+    fetch_var_name='fetch',
+    job_desc=JobDesc(
```

Comment:
Maybe we want to separate job creation and `run`, since `run` can be run multiple times on a same job (e.g., multiple mini-batch steps).

Reply:
`run` call will probably do both submitting the job to k8s cluster and fetch output variables. The problem is submitting job will fail if the job already exists, and `run` may be called multiple times to fetch output. Adding a new interface will cause cluster training code become much different to local training. 

Is it fine to do job submission when creating `RemoteExecutor`?

Reply:
We probably need to support many actions for a job (e.g., creation, view state, deletion), maybe we can use CLI to manage resource, the CLI returns a job ID and the Python script can use the job ID in `run`?

Reply:
OR implement like below is more reasonable? The python program is running on every node, and the k8s job is still managed by CLI which is the same as current implementation?

```python
job_desc = JobDesc(num_trainer=4, num_ps=2, GPU_per_trainer=2)
pserver = RemoteExecutor(placement="/pserver/*", job_desc=job_desc)
trainer = RemoteExecutor(placement="/trainer/*", job_desc=job_desc)

placement = os.getenv('PADDLE_PLACEMENT')
if placement.startswith('/pserver'):
    pserver.run(startup_program)
    pserver.run(optimize_program)
    pserver.wait()
elif placement.startswith('/trainer'):
    for pass_num in xrange(num_passes):
        trainer.run(main_program)
        trainer.wait()    # wait for sync SGD
```

Reply:
Please refer to [this graph](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/refactor/distributed_architecture.md#distributed-training-architecture), there is no Python instance running in the cluster, the python only runs locally, and controls the cpp instances in the cluster.

Optionally we can run a single Python instance in the cluster that controls other cpp instances in the cluster.

The Python code send the entire graph to a service process that will do the graph conversion, placement and send sub graph to each runtime instance. So I think the Python code that the user writes should not contain the placement logic below:

```Python
placement = os.getenv('PADDLE_PLACEMENT')
if placement.startswith('/pserver'):
    pserver.run(startup_program)
    pserver.run(optimize_program)
    pserver.wait()
elif placement.startswith('/trainer'):
    for pass_num in xrange(num_passes):
        trainer.run(main_program)
        trainer.wait()    # wait for sync SGD
```

And only one executor is necessary. So maybe something like:
```Python
job = paddle.createJob(num_trainer=4, num_ps=2, GPU_per_trainer=2) # maybe through a CLI is easier.
exe = paddle.RemoteExecutor(job)
exe.run()
```

Please see [Limitation 1](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/refactor/distributed_architecture.md#limitation-1) on why multiple Python instances in a job will cause problems.

Reply:
1. agree that local python should only do IR creation and submission.
3. I prefer not to differentiate local training job and remote training job, the only difference to the user should be the URI of computing resource(device).
4. And internally, we should not differentiate remote/local devices either.
5. I'm proposing a new class Orchestrator, which is more or less "RemoteExecutor" as mentioned by Helin, the naming is just to hide the idea that where it lives.
6. To unify local or remote training, we might want to treat all training "Remote", so that "local" training is just a special case that resources(devices) are labeled with uri "localhost"
7. To make this happen, Orchestrator, Executors should all expose gRPC or web services to accept ProgramDesc and report back with task status.
8. the whole IR should be submitted to a "Orchestrator", which see the whole picture of the resource it has (local/remote executors), do graph partitioning and send sub IRs to executors(locally or remotely) base on the affiliation config (placement strategy)
9. Orchestrator should be a gRPC service, which is created with executors list and executors affiliation like below:
``` python
orchestrator = paddle.Orchestrator(uri="machine_a.cluster:2222") 
orchestrator.addExecutors(uri="machine_a.cluster:3333")  # add a executor. orchestrator will talk to Executor and fetch device type(CPU, GPU) and other device data
orchestrator.addExecutors(uri="machine_b.cluster:3333")  # add another remote Executor
```
10. then the job will be started with following:
``` python
job_desc = paddle.createJob(num_trainer=4, num_ps=2) # there should be other config regarding affiliation. 
# also, should we wrap job and main_program together? I'm not sure
orchestrator.run(main_program, job_desc) # submit the main_program and job_desc to remote orchestrator
```
if we just need to run local training job, create an orchestrator in localhost and add localhost Executors
11. the benefit is flexibility and separation of concern.
12. when we need to deploy paddle in k8s, it is not necessary to have a designated "master", since any node can be "master", pods may start with an arbitrary sequence. the first one becomes the orchestrator, when new pods are available, just add addExecutors, executors will check what kind of hardware it is assigned, talk to the orchestrator about its hardware feature. this kind of service discovery logic can be done via a k8s adaptor, which watches the pods for this job and call "addExecutor" or "removeExecutor".
13. I know this comment is too long to read, so thanks for reading till the end :P

Reply:
just discussed with @helinwang here is our ideas we want to talk to you with
1. for now we are going to focus on how we are going to launch training with bare metals since this is the fundamental of cluster training.
2. there will be 2 major roles in cluster, Orchestrator and Executor
3. To launch a Orchestrator in a cluster node
``` bash
./orchestrator -port 8080 -ip 192.168.1.2
```
now orchestrator will start listening for incoming resource registration and protobuf
4. to launch an Executor in a cluster node
``` bash
./executor -orchestrator_uri 192.168.1.2:8080
```
now executor is created and register itself with orchestrator with its hardware features
5. to start a cluster training
``` bash
python train.py 192.168.1.2:8080
```
now protobuf (IR) will be sent to the orchestrator; orchestrator change/cut the IR and set sub IR to executors. base on the hardware features of executors, they will be set to different roles(pserver or trainer)

6. to start a local training
``` bash
python train.py
```
IR will be delivered to a lite orchestrator which only do graph change to fit multi thread or multi GPU case, then the updated graph will be delivered to executors and run.

Reply:
Since we are only considering Kubernetes, the `orchestrator` is Kubernetes controller and scheduler; and the `executor` can be the entry of the job image.